### PR TITLE
Popover: refactor to TypeScript

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Part of #35744
Part of #42770

Refactor the `Popover` component to TypeScript.

Unit tests will be converted separately, so that they can offer a better guarantee that the changes in this PR won't be introducing regressions.

## Why?

See #35744 

## How?

Following the step-by-step guide in the [CONTRIBUTING guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#refactoring-a-component-to-typescript)

Reviewing commit-by-commit may also offer an easier way to review this PR.

## Testing Instructions

- Project builds, all tests pass
- For Popover and its derived components (Autocomplete, BorderControl, BorderBoxControl, Dropdown, DropdownMenu, PaletteEdit, GradientPicker, Tooltip..), check Storybook and make sure that all examples behave like on `trunk`
- Do an extensive regression test in the post and site editor

### CODE REVIEW

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:69]: Runtime change: the component was not receiving any props (and it's internal to this file)

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:145]: Runtime change: always return a `motion.div` — I made this change because it would have been otherwise difficult to type `AnimatedWrapper`, since standard HTML elements and motion elements have different types for the same prop names

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:200]: Using new variables because `flip` and `resize` are readonly (`const`)

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:304]: Runtime change: using `firstElementChild` instead of `firstChild`, since `firstChild` can return a `Node` (that doesn't have the `style` property). The additional `instanceof HTMLElement` makes TypeScript happy and adds an extra layer of precaution.

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:319]: This was somehow necessary to make TypeScript happy

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:341]: 😅 

In order to improve types here, we'd need to first assess exactly why `Popover` needs two | three separate callbacks for when the popover closes (one for when the popover is closed by clicking on the `close` button, and one for when the `popover` is closed by clicking outside of the popover).

We'd then need to come up with a good solution that ideally uses only one callback, refactor all usages to the new solution, deprecate this old solution, and then finally we'll be able to remove `onFocusOutside` and the `__unstableOnClose` option from `useDialog`

We should also have a look at [`floating-ui`'s DOM interaction utils](https://floating-ui.com/docs/react-dom-interactions)

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:359]: The default `{}` value was unnecessary and it was throwing off TypeScript

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:424]: Runtime change: moved these checks to be inline because TypeScript wouldn't otherwise use them correctly as a typeguard for `defaultView` later in the code

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx]: I need to potentially look more into why TypeScript doesn't seem to pick up the correct return value of `useDialog`

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:473]: Runtime change: `x` could have still been `null` here, which is not a valid value for style props.

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:515]: Runtime change: somehow `Number.isFinite( arrowData.x )` was not guarding against `arrowData.x` not being `undefined`, and so I added an explicit check (same for `arrowData.y`)

File: packages/components/src/popover/index.tsx
Comment (ciampo) [inline packages/components/src/popover/index.tsx:580]: An alternative to this could be to:

- move the `Popover` component from `index.tsx` to `component.tsx`
- use `index.tsx` to export the `Slot`, `__unstableSlotNameProvider` and `default` exports — something like:

```tsx
export function Slot (
	{ name = SLOT_NAME }: { name?: string },
	ref: ForwardedRef< any >
) {
	return (
		<Slot
			// @ts-expect-error Need to type `SlotFill`
			bubblesVirtually
			name={ name }
			className="popover-slot"
	);
}

export const __unstableSlotNameProvider = slotNameContext.Provider;

export {Popover as default} from './component.tsx`;
```


Thoughts? Could it work?

File: packages/components/src/popover/types.ts
Comment (ciampo) [inline packages/components/src/popover/types.ts]: Not sure if we should even bother here, since currently the README only mentions this format, and we're going to migrate all usage to use `placement` soon (and deprecate `position`)

File: packages/components/src/popover/utils.ts
Comment (ciampo) [inline packages/components/src/popover/utils.ts:168]: Runtime change: function return type changed to `null` instead of `undefined`, in order to match `@floating-ui`'s `setReference` function signature (used in the `Popover` component)

File: packages/components/src/popover/utils.ts
Comment (ciampo) [inline packages/components/src/popover/utils.ts:225]: Runtime change: using `parentElement` instead of `parentNode`, since `Node`s don't have the `getBoundingClientRect()` function and therefore can't be used as the popover's reference (anchor)

**Review by ciampo — COMMENTED** [review]
Hey folks, I'm not completely done yet but this PR is already in a decent state where it wouldn't hurt if you could have a first look and start leaving feedback

cc @mirka @talldan @ntsekouras @walbo @aaronrobertshaw @andrewserong

File: packages/components/src/popover/types.ts
Comment (ciampo) [inline packages/components/src/popover/types.ts:172]: I need to take a look at #40740 to see how this prop was previously used.

Comment (ciampo) [reply packages/components/src/popover/types.ts:172]: It doesn't look like this prop has been used for a long time, will reflect that in its description

File: packages/components/src/popover/index.tsx
Comment (ciampo) [reply packages/components/src/popover/index.tsx]: I decided to type the `useDialog` hook, which removed the need for this `@ts-expect-error` comment

File: packages/components/src/popover/README.md
Comment (aaronrobertshaw) [inline packages/components/src/popover/README.md]:
```suggestion
By default, the _first tabbable element_ in the popover will receive focus when it mounts. This is the same as setting this prop to `"firstElement"`.
```

Typos.

File: packages/components/src/popover/README.md
Comment (aaronrobertshaw) [inline packages/components/src/popover/README.md]:
```suggestion
### `onFocusOutside`: `( event: SyntheticEvent ) => void`
```
Should we match `types.ts` here and explicitly add `void`?

File: packages/components/src/popover/README.md
Comment (aaronrobertshaw) [inline packages/components/src/popover/README.md]:
```suggestion
### `getAnchorRect`: `( fallbackReferenceElement: Element | null ) => DomRectWithOwnerDocument
```

Should this match `types.ts`?

File: packages/components/src/popover/README.md
Comment (aaronrobertshaw) [inline packages/components/src/popover/README.md]:
```suggestion
### `anchorRef`: `Element | PopoverAnchorRefReference | PopoverAnchorRefTopBottom | Range`
```

Should this match `types.ts`?

File: packages/components/src/popover/README.md
Comment (aaronrobertshaw) [inline packages/components/src/popover/README.md]: The type for `children` in `PopoverProps` doesn't explicitly state it is optional. Should this be `Yes`?

File: packages/components/src/popover/types.ts
Comment (aaronrobertshaw) [inline packages/components/src/popover/types.ts]:
```suggestion
	 * By default, the _first tabbable element_ in the popover will receive focus
```

Typo.

File: packages/components/src/popover/README.md
Comment (aaronrobertshaw) [inline packages/components/src/popover/README.md]:
```suggestion
```

There's no `@default` for `isAlternate` in `types.ts`. This also doesn't appear to match the `boolean` type. Should this be here?

File: packages/components/src/popover/types.ts
Comment (aaronrobertshaw) [inline packages/components/src/popover/types.ts:172]: Should this `range` prop also be included in the README and noted as deprecated there along with the other deprecated props such as `__unstableShift`?

File: packages/components/src/popover/README.md
Comment (aaronrobertshaw) [inline packages/components/src/popover/README.md]: This prop has been flagged as deprecated in `types.ts` should we note that here as well?

File: packages/components/src/popover/types.ts
Comment (aaronrobertshaw) [inline packages/components/src/popover/types.ts:158]: Should this prop also be included in the README?

`__unstableShift` for example is included despite it being both unstable and deprecated.

**Review by aaronrobertshaw — COMMENTED** [review]
Great work as always @ciampo 👍 

I've only had a chance to give this an initial read-through and test in Storybook, but so far it's testing well.

✅ No build or typing errors
✅ Unit tests pass for BorderControl, BorderBoxControl, and Popover.
✅ Didn't notice any inconsistent behaviour between this PR and trunk in Storybook. Tested Popover, border controls and others derived from the Popover e.g. Dropdown.

❓ I did spot a few typos and inconsistencies between the `types.ts` file and the Popover readme. I've left some inline comments for those.

I'll aim to do some more testing within the block and site editors in the next few days unless others beat me to it.

File: packages/components/src/popover/README.md
Comment (ciampo) [reply packages/components/src/popover/README.md]: https://github.com/WordPress/gutenberg/pull/43823/commits/5ef6066306805d191df3a0bdbc7b46d594e8cbc3

File: packages/components/src/popover/types.ts
Comment (ciampo) [reply packages/components/src/popover/types.ts]: https://github.com/WordPress/gutenberg/pull/43823/commits/5ef6066306805d191df3a0bdbc7b46d594e8cbc3

File: packages/components/src/popover/README.md
Comment (ciampo) [reply packages/components/src/popover/README.md]: https://github.com/WordPress/gutenberg/pull/43823/commits/f9f1f5d49e076fe508c77184885d19512c5e9e4d

File: packages/components/src/popover/README.md
Comment (ciampo) [reply packages/components/src/popover/README.md]: https://github.com/WordPress/gutenberg/pull/43823/commits/1d9b8a2a4dff16c698404a3e2891ad969ba4f6fb

File: packages/components/src/popover/README.md
Comment (ciampo) [reply packages/components/src/popover/README.md]: https://github.com/WordPress/gutenberg/pull/43823/commits/8af91037d9b9290cb75e2908bb1e91d25a327dc3

File: packages/components/src/popover/README.md
Comment (ciampo) [reply packages/components/src/popover/README.md]: https://github.com/WordPress/gutenberg/pull/43823/commits/828cb7469d4a0b7f07d0e301d72b65fee5e49ad1

File: packages/components/src/popover/README.md
Comment (ciampo) [reply packages/components/src/popover/README.md]: https://github.com/WordPress/gutenberg/pull/43823/commits/67d8fe7db0dff0126fdab19912b1299dc2213ffb

File: packages/components/src/popover/types.ts
Comment (ciampo) [reply packages/components/src/popover/types.ts:172]: Good point. I decided to remove all `__unstable` props from the README, and added the `range` prop instead https://github.com/WordPress/gutenberg/pull/43823/commits/57692b878ae48de0ecc6133fadeb6e02faf321bd

File: packages/components/src/popover/README.md
Comment (ciampo) [reply packages/components/src/popover/README.md]: Good point. I decided to remove all `__unstable` props from the README, and added the `range` prop instead https://github.com/WordPress/gutenberg/pull/43823/commits/57692b878ae48de0ecc6133fadeb6e02faf321bd

File: packages/components/src/popover/types.ts
Comment (ciampo) [reply packages/components/src/popover/types.ts:158]: Good point. I decided to remove all `__unstable` props from the README, and added the `range` prop instead https://github.com/WordPress/gutenberg/pull/43823/commits/57692b878ae48de0ecc6133fadeb6e02faf321bd

**Comment by ciampo** [comment]
Thank you @aaronrobertshaw ! I missed your reviews :) 

>❓ I did spot a few typos and inconsistencies between the types.ts file and the Popover readme. I've left some inline comments for those.

Thank you for pointing those out! The README should be in a much better place now

> I'll aim to do some more testing within the block and site editors in the next few days unless others beat me to it.

🙇

File: packages/components/src/border-control/border-control-dropdown/component.tsx
Comment (mirka) [inline packages/components/src/border-control/border-control-dropdown/component.tsx]: Just checking: The reasoning behind this change is that consumers should be responsible for changing the object reference when necessary?

File: packages/components/src/popover/stories/index.tsx
Comment (mirka) [inline packages/components/src/popover/stories/index.tsx]: Should we hide the code snippets for this one?

We can also remove the `@example` tag in the main JSDoc.

File: packages/components/src/popover/stories/index.tsx
Comment (mirka) [inline packages/components/src/popover/stories/index.tsx:53]: Shouldn't `children` still be disabled?

**Review by mirka — APPROVED** [review]
Nice! I did a smoke test of all the listed components in the editor, and nothing seemed off.

The only thing I noticed was that the color popover of PaletteEdit has a different offset compared to the color popovers in edit-post. (In edit-post, the popover doesn't overlap the writing flow scroll bar.) But this was not a regression in this PR, so 👍 

image

File: packages/components/src/popover/README.md
Comment (ciampo) [inline packages/components/src/popover/README.md:96]: Removed `container` as a possible option as it looks like it was removed in https://github.com/WordPress/gutenberg/pull/27699

File: packages/components/src/border-control/border-control-dropdown/component.tsx
Comment (ciampo) [reply packages/components/src/border-control/border-control-dropdown/component.tsx]: No, I simply wanted to simplify the code and avoid the creation of a new object every time — was that intended? Happy to revert

**Comment by ciampo** [comment]
> The only thing I noticed was that the color popover of PaletteEdit has a different offset compared to the color popovers in edit-post. (In edit-post, the popover doesn't overlap the writing flow scroll bar.) But this was not a regression in this PR, so 👍

I'll try to look into it after this PR is merged 👌

File: packages/components/src/popover/stories/index.tsx
Comment (ciampo) [reply packages/components/src/popover/stories/index.tsx:53]: https://github.com/WordPress/gutenberg/pull/43823/commits/25d4391dceaf31d58d412bd82794833402199eea

File: packages/components/src/popover/stories/index.tsx
Comment (ciampo) [reply packages/components/src/popover/stories/index.tsx]: https://github.com/WordPress/gutenberg/pull/43823/commits/78a25e8d453ff3252f25a13ba2b6b585be0960c1

File: packages/components/src/border-control/border-control-dropdown/component.tsx
Comment (mirka) [reply packages/components/src/border-control/border-control-dropdown/component.tsx]: Seems intentional to me, although it's unclear whether it's warranted. If you haven't looked into it specifically, it might be better to bump that change to a separate issue, since it seems like the kind of thing that would have *implications* 👻 (It might be good implications though!)

Comment (ciampo) [reply packages/components/src/border-control/border-control-dropdown/component.tsx]: The last PR to edit those lines was #41900

Anyway, I went ahead and reverted the above change in https://github.com/WordPress/gutenberg/pull/43823/commits/dac81ea2a8b33f5e9d28d17a0491db8c70adb93c

**Comment by ciampo** [comment]
All feedback has been addressed — I will merge once CI is green ✅

**Pull request merged**
